test: Gsheet Automation

sarojsarab opened the pull request:
## Description
- Automated below queries for gsheet

1. Fetch Details
2. Insert one
3. Insert many
4. Update one
5. Update many
6. Fetch many
7. Delete One

- Added workflow to schedule run for gsheet automated tests on the hosted instance
- Added a new cypress config file which will be used to run the gsheet tests
- Added tests for folowing permission/scope options  `All access`, `selected access`, `Read/write | All google sheets` and `Read | All google sheets`
- Added negative scenarios
- Added tests to verify widget binding for both suggested widget and drag n drop widget for both selected and all access permission/scope
#### Type of change
- Cypress
## Testing
>
#### How Has This Been Tested?
- [x] Cypress
>
>
## Checklist:

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

Aishwarya-U-R reviewed (commented):
Can you update PR description to include the scenarios which are tested in this PR?
And where is sheet exactly authenticated?

sarojsarab commented:
> Can you update PR description to include the scenarios which are tested in this PR?
> And where is sheet exactly authenticated?

Thanks for the feedback, I have updated the description with what has been covered.

To answer your 2nd question, we are not authenticating gsheet from code as it is not possible, we have hosted appsmith at https://regression.test.appsmith.com and have an existing workspace and authenticated gsheet datasource to run the tests. We will not be running gsheet tests as part of our regular CI runs, I have added a new workflow to run the gsheet tests on a schedule basis.

Aishwarya-U-R commented on app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts:
```diff
@@ -0,0 +1,193 @@
+
+    //Adding query to insert a new spreadsheet
+    homePage.NavigateToHome();
+    homePage.CreateAppInWorkspace(workspaceName);
```
If a workspace in that machine is already have workspace with sheet authorised, why are we creating a new workspace here?

sarojsarab replied:
We are not creating workspace here, we are creating a new app here in the existing workspace.

Aishwarya-U-R commented on app/client/cypress/support/Pages/DataSources.ts:
```diff
@@ -1534,4 +1539,40 @@ export class DataSources {
     });
   }
+
+  public enterSortByValues(sortBy: string, option: string, index = 0) {
```
TS helper methods names are CamelCase - `EnterSortByValues`

sarojsarab replied:
Fixed. Thanks

Aishwarya-U-R commented on app/client/cypress/fixtures/test-data-gsheet.ts:
```diff
@@ -0,0 +1,156 @@
+export const GSHEET_DATA = [
```
What is use of this file?

sarojsarab replied:
We use this data for insert/update to the spreadsheet

sarojsarab commented on .github/workflows/ci-test-gsheet.yml:
```diff
@@ -0,0 +1,443 @@
+name: Appsmith CI Gsheet Test Workflow
```
Updated, thanks!

Aishwarya-U-R commented on app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts:
```diff
@@ -0,0 +1,193 @@
+
+    // Sort by descending and verify
+    dataSources.clearSortByOption(); //clearing previous sort option
+    dataSources.enterSortByValues("price", "Descending");
```
References should also be corrected.

sarojsarab replied:
Corrected the references, thanks!

sarojsarab commented:
/ok-to-test

sarojsarab commented:
/ok-to-test

Aishwarya-U-R commented on app/client/cypress/e2e/GSheet/GSheetWidgetBinding_spec.ts:
```diff
@@ -0,0 +1,87 @@
+/// <reference types="Cypress" />
```
Is this line needed?

sarojsarab replied:
I think it's better to be there as is used for type checking and intellisense in Cypress tests for typescript, also it helps the TypeScript compiler to understand the Cypress-specific syntax and provides better code suggestions and error checking.

Aishwarya-U-R reviewed (dismissed):
Approving current changes.
Can we merge this PR & start running @sarojsarab ? Any new tests from Prapula can come in later as new PR.

sarojsarab commented:
> Approving current changes.
> Can we merge this PR & start running @sarojsarab ? Any new tests from Prapula can come in later as new PR.
Thanks @Aishwarya-U-R 
And yes, let's merge this. I will raise new PR for upcoming changes.

Aishwarya-U-R commented on app/client/cypress/support/Pages/DataSources.ts:
```diff
@@ -1,6 +1,7 @@
 import { ObjectsRegistry } from "../Objects/Registry";
 import { WIDGET } from "../../locators/WidgetLocators";
 import { EntityItems } from "./AssertHelper";
+import formControl from "../../locators/FormControl.json";
```
Bring the needed locator from formControl to this file & remove this import..

sarojsarab replied:
Done!

Aishwarya-U-R commented on app/client/cypress/support/Pages/DataSources.ts:
```diff
@@ -231,10 +233,15 @@ export class DataSources {
     ".datasourceStructure-query-editor";
   private _datasourceColumnSchemaInQueryEditor = ".t--datasource-column";
   private _datasourceStructureSearchInput = ".datasourceStructure-search input";
+  _JSModeSortingControl = ".t--actionConfiguration\\.formData\\.sortBy\\.data";
```
Variable name starts small case` _jsModeSortingControl`

sarojsarab replied:
Done!

sarojsarab commented:
/ok-to-test

sarojsarab commented:
/ok-to-test

sarojsarab commented:
/ok-to-test
